### PR TITLE
Add first functional test for the search service

### PR DESCRIPTION
**Pull request by joelverhagen**
1. Add stitch together code that downloads packages, builds a Lucene index, and starts the server.
2. Add simple functional test for NuGet/Home#833.

Please give me any feedback you can on this framework. I am still pretty ignorant about this stuff. Areas in particular that need review:

- Should we download packages from nuget.org? I am `DataServicePackageRepository` to do this. Alternatives are reading .nupkgs from a share or checking them in maybe?
- Should each test start and stop the service on its own? Right now I use a class fixture to do all of the setup once per *test run*. That is, all tests share the same service, Lucene index, and data directory.
- Are the directories for the test data okay? I am using `Directory.GetCurrentDirectory()` as the base. I think TeamCity cleans this up automatically.
- Should I explicitly clean up the test data? Always? On successful tests (this one may be tricky)?

@johnataylor @joyhui @maartenba @xavierdecoster @yishaigalatzer

### CODE REVIEW

File: src/NuGet.Services.BasicSearch/Startup.cs [hunk 1]
```diff
@@ -171,16 +171,16 @@ static Uri MakeRegistrationBaseAddress(string scheme, string registrationBaseAdd
             {
                 if (_searcherManager == null)
                 {
-                    await context.Response.WriteAsync("UNINITIALIZED");
                     context.Response.StatusCode = (int)HttpStatusCode.OK;
+                    await context.Response.WriteAsync("UNINITIALIZED");
```
Comment (joelverhagen) [inline src/NuGet.Services.BasicSearch/Startup.cs]: Response needs to be written after status code is sent.

File: tests/NuGet.Services.BasicSearchTests/StartupTests.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    {
+        private readonly StartupTestsFixture _fixture;
+
+        public StartupTests(StartupTestsFixture fixture)
+        {
+            _fixture = fixture;
+        }
+
+        [Fact]
+        public async Task AutocompleteParameterIsObserved()
```
Comment (joelverhagen) [inline tests/NuGet.Services.BasicSearchTests/StartupTests.cs]: This is the one test.

File: tests/NuGet.Services.BasicSearchTests/TestSupport/DataInitializer.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+            _settings = settings;
+        }
+
+        public async Task InitializeAsync()
+        {
+            Directory.CreateDirectory(_settings.DataDirectory);
+            await WriteFileAsync("downloads.v1.json", "[]");
+            await WriteFileAsync("curatedfeeds.json", "[]");
+            await WriteFileAsync("owners.json", "[]");
+            await WriteFileAsync("rankings.v1.json", "{\"Rank\": []}");
```
Comment (joelverhagen) [inline tests/NuGet.Services.BasicSearchTests/TestSupport/DataInitializer.cs]: This essentially makes search ranking by download count ineffective. Future tests may want to test this.

File: tests/NuGet.Services.BasicSearchTests/Fixtures/StartupTestsFixture.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+                DataDirectory = GetDirectoryPath("Data"),
+                RegistrationBaseAddress = "http://api.nuget.org/v3/registration0/"
+            };
+
+            // initialize the Lucene and Data directories
+            await new LuceneInitializer(settings).InitializeAsync();
+            await new DataInitializer(settings).InitializeAsync();
+
+            // establish configuration for the web service
+            ConfigurationManager.AppSettings.Set("Local.Lucene.Directory", settings.LuceneDirectory);
```
Comment (joelverhagen) [inline tests/NuGet.Services.BasicSearchTests/Fixtures/StartupTestsFixture.cs]: This is intrinsically not thread safe. Therefore, if we ever want different tests having different settings, we will need to refactor the service.

File: tests/NuGet.Services.BasicSearchTests/TestSupport/PackageVersionIdentifier.cs [hunk 5]
```diff
@@ -0,0 +1,50 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace NuGet.Services.BasicSearchTests.TestSupport
+{
+    public class PackageVersionIdentifier
```
Comment (joyhui) [inline tests/NuGet.Services.BasicSearchTests/TestSupport/PackageVersionIdentifier.cs]: It is typical example for a "struct".

File: tests/NuGet.Services.BasicSearchTests/TestSupport/LuceneInitializer.cs [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+        private async Task DownloadPackagesAsync()
+        {
+            System.IO.Directory.CreateDirectory(_settings.PackageDirectory);
+            await Task.WhenAll(_settings.PackageVersions.Select(DownloadPackageAsync));
+        }
+
+        private void CreateLuceneIndex()
+        {
+            var directoryInfo = new DirectoryInfo(_settings.LuceneDirectory);
+            directoryInfo.Create();
```
Comment (joyhui) [inline tests/NuGet.Services.BasicSearchTests/TestSupport/LuceneInitializer.cs]: If the lucene index already created, do we clear old one or reuse old one?
People may run function test on his box more than once.

Comment (joelverhagen) [inline tests/NuGet.Services.BasicSearchTests/TestSupport/LuceneInitializer.cs]: Yes, the `true` parameter to `Document.CreateIndexWriter` does this.

File: tests/NuGet.Services.BasicSearchTests/TestSupport/DataInitializer.cs (hunk 3 above)
Comment (joyhui) [inline tests/NuGet.Services.BasicSearchTests/TestSupport/DataInitializer.cs]: We may have to expose interface to change/initiate those file too. 
One idea is to add different section to the BasicSearchTestSettings itself, or let individual test to set content by themselves using interface.

Comment (joelverhagen) [inline tests/NuGet.Services.BasicSearchTests/TestSupport/DataInitializer.cs]: Yup, that's one option. I was also thinking that it could be set as part of the list of packages to download and include in the Lucene index (that is, the `PackageVersionIdentifier` type). If it's okay with you, I'll wait until we need it to build it.

File: tests/NuGet.Services.BasicSearchTests/TestSupport/PackageVersionIdentifier.cs (hunk 5 above)
Comment (joelverhagen) [inline tests/NuGet.Services.BasicSearchTests/TestSupport/PackageVersionIdentifier.cs]: Good point. Fixed.

File: tests/NuGet.Services.BasicSearchTests/Fixtures/StartupTestsFixture.cs [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+                        new PackageVersionIdentifier("Antlr", "3.1.3.42154"),
+                        new PackageVersionIdentifier("Antlr", "3.4.1.9004-pre"),
+                        new PackageVersionIdentifier("angularjs", "1.2.0-RC1"),
+                        new PackageVersionIdentifier("WebGrease", "1.6.0")
+                    }
+                }
+            };
+
+            // construct the test settings
+            var settings = new BasicSearchTestSettings
```
Comment (joyhui) [inline tests/NuGet.Services.BasicSearchTests/Fixtures/StartupTestsFixture.cs]: Please use a xml file or json file for this.

File: tests/NuGet.Services.BasicSearchTests/StartupTests.cs [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+            Assert.Equal("1.2.0-RC1", withPrerelease.GetPackageVersion("angularjs"));
+            Assert.Equal("3.4.1.9004-pre", withPrerelease.GetPackageVersion("Antlr"));
+            Assert.Equal("1.6.0", withPrerelease.GetPackageVersion("WebGrease"));
+
+            Assert.False(withoutPrerelease.ContainsPackage("angularjs"));
+            Assert.Equal("3.1.3.42154", withoutPrerelease.GetPackageVersion("Antlr"));
+            Assert.Equal("1.6.0", withoutPrerelease.GetPackageVersion("WebGrease"));
+        }
+
+        private static string Query(object obj)
```
Comment (joyhui) [inline tests/NuGet.Services.BasicSearchTests/StartupTests.cs]: Since the format of query for search is fixed, can we add a helper class for it?

class SearchQueryBuilder
{
    public static string base; --- static or retrieve from somewhere else 
    public int skip {get;set}
    public bool prerease {get;set}
    public override ToString() ---//  base?query?.....
}

File: tests/NuGet.Services.BasicSearchTests/StartupTests.cs [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+            var withoutPrereleaseResponse = await app.Client.GetAsync("/query?" + Query(new { q = query, prerelease = false }));
+            var withoutPrerelease = await withoutPrereleaseResponse.Content.ReadAsAsync<QueryResult>();
+
+            // Assert
+            Assert.Equal(HttpStatusCode.OK, withPrereleaseResponse.StatusCode);
+            Assert.Equal(HttpStatusCode.OK, withoutPrereleaseResponse.StatusCode);
+
+            Assert.Equal("1.2.0-RC1", withPrerelease.GetPackageVersion("angularjs"));
+            Assert.Equal("3.4.1.9004-pre", withPrerelease.GetPackageVersion("Antlr"));
```
Comment (joyhui) [inline tests/NuGet.Services.BasicSearchTests/StartupTests.cs]: It is good always put an failed msg at the end:
Assert.Equal("3.4.1.9004-pre", withPrerelease.GetPackageVersion("Antlr"), "Antlr package should be returned");

File: tests/NuGet.Services.BasicSearchTests/StartupTests.cs [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+namespace NuGet.Services.BasicSearchTests
+{
+    public class StartupTests : IClassFixture<StartupTestsFixture>
+    {
+        private readonly StartupTestsFixture _fixture;
+
+        public StartupTests(StartupTestsFixture fixture)
+        {
+            _fixture = fixture;
+        }
```
Comment (joyhui) [inline tests/NuGet.Services.BasicSearchTests/StartupTests.cs]: We may want to set the packageVersion at startup, build lucene, and let service refresh. In this way we could make sure lucene is clean and won't affect other tests.

File: tests/NuGet.Services.BasicSearchTests/StartupTests.cs (hunk 9 above)
Comment (joelverhagen) [inline tests/NuGet.Services.BasicSearchTests/StartupTests.cs]: I was wanting to do this too, but Xunit does not support this (except on `Assert.Fail`).

File: tests/NuGet.Services.BasicSearchTests/StartupTests.cs (hunk 10 above)
Comment (joelverhagen) [inline tests/NuGet.Services.BasicSearchTests/StartupTests.cs]: I think you're right. I'll do this.

File: tests/NuGet.Services.BasicSearchTests/Fixtures/StartupTestsFixture.cs (hunk 4 above)
Comment (maartenba) [inline tests/NuGet.Services.BasicSearchTests/Fixtures/StartupTestsFixture.cs]: My vote is on using the approach we have in the gallery functional tests: use a `ConfigurationProvider` that does a fallback:

* Environment variable with that name set? Use environment variable.
* No environment variable but a setting exists? Use setting.
* None set? `null`

This makes it easier to change the test setting during a run. Optionally a test setup/teardown can set/unset a variable as well.

File: tests/NuGet.Services.BasicSearchTests/TestSupport/LuceneInitializer.cs (hunk 6 above)
Comment (maartenba) [inline tests/NuGet.Services.BasicSearchTests/TestSupport/LuceneInitializer.cs]: Why not a simple `RAMDirectory` here? Wondering if each test(class) should just create a fresh index for it to run. Makes the tests very isolated and less prone to environmental issues (like a `write.lock` that remains and other things that can go wrong with the file system)

File: tests/NuGet.Services.BasicSearchTests/Fixtures/StartupTestsFixture.cs (hunk 7 above)
Comment (maartenba) [inline tests/NuGet.Services.BasicSearchTests/Fixtures/StartupTestsFixture.cs]: If it's only four of them I prefer code. Makes me grasp the test when reading it without having to context switch to yet another file.

File: tests/NuGet.Services.BasicSearchTests/StartupTests.cs (hunk 9 above)
Comment (maartenba) [inline tests/NuGet.Services.BasicSearchTests/StartupTests.cs]: `Assert` can be extended but I don't think we should write our own xUnit. Perhaps add a line comment describing the assertion so that when you navigate to the failed assertion some extra info is available to the dev?

**Comment by maartenba** [comment]
In reply to your questions:

> Should we download packages from nuget.org? I am DataServicePackageRepository to do this. Alternatives are reading .nupkgs from a share or checking them in maybe?

No strong opinion. Since NuGet.org has the premise of "no deletes", it should be safe to depend on it to fetch packages.

> Should each test start and stop the service on its own? Right now I use a class fixture to do all of the setup once per test run. That is, all tests share the same service, Lucene index, and data directory.

My vote is on service per test. Makes them all isolated and very much able to run in parallel. If we expect a delete in some test, for example, a shared service means the test can not run in parallel if another test depends on the deleted package being there.

> Are the directories for the test data okay? I am using Directory.GetCurrentDirectory() as the base. I think TeamCity cleans this up automatically.
Should I explicitly clean up the test data? Always? On successful tests (this one may be tricky)?

How about a `RAMDirectory`?

File: src/NuGet.Services.BasicSearch/Startup.cs [hunk 11]
```diff
@@ -204,8 +204,8 @@ static Uri MakeRegistrationBaseAddress(string scheme, string registrationBaseAdd
                             break;
 
                         default:
-                            await context.Response.WriteAsync("unrecognized");
                             context.Response.StatusCode = (int)HttpStatusCode.NotFound;
+                            await context.Response.WriteAsync("unrecognized");
```
Comment (maartenba) [inline src/NuGet.Services.BasicSearch/Startup.cs]: Other messages are in all caps.

File: tests/NuGet.Services.BasicSearchTests/TestSupport/DataInitializer.cs (hunk 3 above)
Comment (maartenba) [inline tests/NuGet.Services.BasicSearchTests/TestSupport/DataInitializer.cs]: Just like the `RAMDirectory` comment, I always feel wary when tests depend on the file system. If we introduce a test that does require this data to be populated, we can no longer run them in parallel (unless we use a different directory per test). Imo the code should take in the dictionary/enumerable, not the file.

File: tests/NuGet.Services.BasicSearchTests/TestSupport/LuceneInitializer.cs [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+            var path = GetPackagePath(version);
+            using (var fileStream = new FileStream(path, FileMode.Open, FileAccess.Read))
+            {
+                var errors = new List<string>();
+                var metadata = PackageMetadataExtraction.MakePackageMetadata(fileStream, errors);
+
+                if (errors.Any())
+                {
+                    throw new InvalidOperationException(
+                        $"NuGet package for '{version.Id}' (version '{version.Version}') could be read for metadata. " +
```
Comment (maartenba) [inline tests/NuGet.Services.BasicSearchTests/TestSupport/LuceneInitializer.cs]: could *not* be read

File: tests/NuGet.Services.BasicSearchTests/TestSupport/LuceneInitializer.cs [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+                return metadata;
+            }
+        }
+
+        private string GetPackagePath(PackageVersionIdentifier version)
+        {
+            return Path.Combine(_settings.PackageDirectory, $"{version.Id}.{version.Version}.nupkg".ToLower());
+        }
+        
+        private async Task DownloadPackageAsync(PackageVersionIdentifier version)
```
Comment (maartenba) [inline tests/NuGet.Services.BasicSearchTests/TestSupport/LuceneInitializer.cs]: Wondering if a `NuGet.Core` dependency is needed. We could just do a download from the DNX flat container?

File: tests/NuGet.Services.BasicSearchTests/Fixtures/StartupTestsFixture.cs (hunk 4 above)
Comment (joelverhagen) [inline tests/NuGet.Services.BasicSearchTests/Fixtures/StartupTestsFixture.cs]: I think I will address this in a different PR. I wanted to minimize changes to the product code to get this functional test in place.

File: tests/NuGet.Services.BasicSearchTests/StartupTests.cs (hunk 9 above)
Comment (joelverhagen) [inline tests/NuGet.Services.BasicSearchTests/StartupTests.cs]: Good compromise. I'll do that.

File: tests/NuGet.Services.BasicSearchTests/TestSupport/LuceneInitializer.cs (hunk 6 above)
Comment (joelverhagen) [inline tests/NuGet.Services.BasicSearchTests/TestSupport/LuceneInitializer.cs]: This will required a lot of changes to the service (since it only uses a file system directory right now). I would like to do this as a separate PR, if that's okay.

Comment (maartenba) [inline tests/NuGet.Services.BasicSearchTests/TestSupport/LuceneInitializer.cs]: Sounds good!

File: tests/NuGet.Services.BasicSearchTests/TestSupport/LuceneInitializer.cs (hunk 13 above)
Comment (joelverhagen) [inline tests/NuGet.Services.BasicSearchTests/TestSupport/LuceneInitializer.cs]: I am not sure what you mean about this. Could you elaborate or point me to some code that does this?

File: tests/NuGet.Services.BasicSearchTests/TestSupport/LuceneInitializer.cs (hunk 12 above)
Comment (joelverhagen) [inline tests/NuGet.Services.BasicSearchTests/TestSupport/LuceneInitializer.cs]: Fixed.

File: src/NuGet.Services.BasicSearch/Startup.cs (hunk 11 above)
Comment (joelverhagen) [inline src/NuGet.Services.BasicSearch/Startup.cs]: Fixed.

File: tests/NuGet.Services.BasicSearchTests/StartupTests.cs (hunk 8 above)
Comment (joelverhagen) [inline tests/NuGet.Services.BasicSearchTests/StartupTests.cs]: Fixed.

File: tests/NuGet.Services.BasicSearchTests/Fixtures/StartupTestsFixture.cs (hunk 7 above)
Comment (joelverhagen) [inline tests/NuGet.Services.BasicSearchTests/Fixtures/StartupTestsFixture.cs]: I've made the package list inline with the code and values the same for all tests in XML.

File: tests/NuGet.Services.BasicSearchTests/TestSupport/LuceneInitializer.cs (hunk 13 above)
Comment (maartenba) [inline tests/NuGet.Services.BasicSearchTests/TestSupport/LuceneInitializer.cs]: If this is the only place `NuGet.Core` is used, you could just HTTP GET the package instead using the following URL format:
`https://api.nuget.org/v3-flatcontainer/{packageid-lowercase}/{packageversion-lowercase}/{packageid-lowercase}.{packageversion-lowercase}.nupkg`

(example `https://api.nuget.org/v3-flatcontainer/newtonsoft.json/4.0.1/newtonsoft.json.4.0.1.nupkg`)

Comment (joelverhagen) [inline tests/NuGet.Services.BasicSearchTests/TestSupport/LuceneInitializer.cs]: Okay, this is a lot simpler. I'll do this.

File: src/NuGet.Indexing/DocumentCreator.cs [hunk 14]
```diff
@@ -264,14 +264,32 @@ static void AddSupportedFrameworks(Document document, IDictionary<string, string
                 }
             }
         }
+        static void AddRequiredField(Document document, string destName, IDictionary<string, string> package, List<string> errors, string sourceName, Field.Index fieldIndex, float boost = 1.0f)
```
Comment (yishaigalatzer) [inline src/NuGet.Indexing/DocumentCreator.cs]: private static void (and newline above this method), also try to break the long lines, so the whole line is readable when reviewing the PR

File: src/NuGet.Indexing/DocumentCreator.cs [hunk 15]
```diff
@@ -264,14 +264,32 @@ static void AddSupportedFrameworks(Document document, IDictionary<string, string
...
+            {
+                document.Add(new Field(destName, defaultValue, Field.Store.YES, fieldIndex) { Boost = boost });
+            }
+        }
 
-        static void AddField(Document document, string destName, IDictionary<string, string> package, string sourceName, Field.Index fieldIndex, float boost = 1.0f)
+        static bool AddField(Document document, string destName, IDictionary<string, string> package, string sourceName, Field.Index fieldIndex, float boost = 1.0f)
         {
             string value;
             if (package.TryGetValue(sourceName, out value))
```
Comment (yishaigalatzer) [inline src/NuGet.Indexing/DocumentCreator.cs]: can the value be accidentially be null or empty?

File: src/NuGet.Indexing/PackageMetadataExtraction.cs [hunk 16]
```diff
@@ -93,6 +94,9 @@ public static void MakePackage(Stream stream, IDictionary<string, string> packag
             ExtractProperty(package, document, "licenseUrl");
             ExtractProperty(package, document, "requiresLicenseAcceptance");
             ExtractDependencies(package, document);
+
+            package["published"] = DateTime.UtcNow.ToString("O");
```
Comment (yishaigalatzer) [inline src/NuGet.Indexing/PackageMetadataExtraction.cs]: can we switch to using DateTimeOffset instead of DateTime. Generally DateTimeOffset is recommended to be used whenever you can instead of DateTime.

File: src/NuGet.Indexing/PackageMetadataExtraction.cs [hunk 17]
```diff
@@ -110,39 +114,24 @@ static bool TryLoad(Stream stream, out XDocument document, List<string> errors)
...
-            else
+            if (!ExtractProperty(package, document, localName, key))
             {
-                errors.Add("unable to find the version element in the nuspec");
+                errors.Add($"unable to find the '{localName}' element in the nuspec");
             }
         }
 
-        static void ExtractProperty(IDictionary<string, string> package, XDocument document, string name)
+        static bool ExtractProperty(IDictionary<string, string> package, XDocument document, string localName, string key = null)
```
Comment (yishaigalatzer) [inline src/NuGet.Indexing/PackageMetadataExtraction.cs]: this is kind of a weird API, if you really want to keep it, please add an doc explaining the usage.

Also in general you want to avoid methods with default parameters values

File: src/NuGet.Services.BasicSearch/Startup.cs (hunk 1 above)
Comment (yishaigalatzer) [inline src/NuGet.Services.BasicSearch/Startup.cs]: :+1: which is why in webapi a response is a single object that travels up the pipe. Such that if any error happens along the way of producing the result, you have a chance to replace the result with a 500 or another proper error that end user can see

File: tests/NuGet.Services.BasicSearchTests/TestSupport/StartedWebApp.cs [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+            ConfigurationManager.AppSettings.Set("Local.Lucene.Directory", luceneDirectory);
+            ConfigurationManager.AppSettings.Set("Local.Data.Directory", dataDirectory);
+            ConfigurationManager.AppSettings.Set("Search.RegistrationBaseAddress", _settings.RegistrationBaseAddress);
+
+            // start the app
+            _webApp = WebApp.Start<Startup>(_portReserver.BaseUri);
+            Client = new System.Net.Http.HttpClient { BaseAddress = new Uri(_portReserver.BaseUri) };
+        }
+
+        public System.Net.Http.HttpClient Client { get; private set; }
```
Comment (yishaigalatzer) [inline tests/NuGet.Services.BasicSearchTests/TestSupport/StartedWebApp.cs]: put namespaces in the using statements on top

File: tests/NuGet.Services.BasicSearchTests/TestSupport/StartedWebApp.cs [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+        private TestSettings _settings;
+        private INupkgDownloader _nupkgDownloader;
+        private LuceneDirectoryInitializer _luceneDirectoryInitializer;
+        private DataDirectoryInitializer _dataDirectoryInitializer;
+        private PortReserver _portReserver;
+        private IDisposable _webApp;
+
+        public static async Task<StartedWebApp> StartAsync(IEnumerable<PackageVersion> packages = null)
+        {
+            var startedWebApp = new StartedWebApp();
```
Comment (yishaigalatzer) [inline tests/NuGet.Services.BasicSearchTests/TestSupport/StartedWebApp.cs]: if you don't want it to be constructed elsewhere, make the constructor private, and the same for initialize async

File: src/NuGet.Indexing/DocumentCreator.cs (hunk 15 above)
Comment (joelverhagen) [inline src/NuGet.Indexing/DocumentCreator.cs]: `null` will be rejected by rejected by Lucene so adding additional validation is not necessary. `String.Empty` is an acceptable input.

File: tests/NuGet.Services.BasicSearchTests/TestSupport/StartedWebApp.cs (hunk 18 above)
Comment (joelverhagen) [inline tests/NuGet.Services.BasicSearchTests/TestSupport/StartedWebApp.cs]: This was necessary when referencing NuGet.Core because it has an HttpClient in the NuGet.* parent namespace. Per @maartenba's comment above, this full qualification is no longer necessary as this dependency has been removed.

File: tests/NuGet.Services.BasicSearchTests/TestSupport/StartedWebApp.cs (hunk 19 above)
Comment (joelverhagen) [inline tests/NuGet.Services.BasicSearchTests/TestSupport/StartedWebApp.cs]: Fixed.

File: src/NuGet.Indexing/PackageMetadataExtraction.cs (hunk 17 above)
Comment (joelverhagen) [inline src/NuGet.Indexing/PackageMetadataExtraction.cs]: Fixed.

File: src/NuGet.Indexing/PackageMetadataExtraction.cs (hunk 16 above)
Comment (joelverhagen) [inline src/NuGet.Indexing/PackageMetadataExtraction.cs]: Fixed.

File: src/NuGet.Indexing/DocumentCreator.cs (hunk 14 above)
Comment (joelverhagen) [inline src/NuGet.Indexing/DocumentCreator.cs]: Fixed.

**Comment by maartenba** [comment]
:shipit:

**Comment by joelverhagen** [comment]
This makes some progress on NuGet/NuGetGallery#2798.
